File: gsd09-web-crawler/web-crawler-app/src/main/java/webcrawler/urldiscovery/service/UrlDiscoveryService.java
package webcrawler.urldiscovery.service;

import webcrawler.urldiscovery.producer.UrlDiscoveryEventProducer;
import webcrawler.urldiscovery.request.UrlDiscoveryBatchResponseDto;
import webcrawler.urldiscovery.request.UrlDiscoveryRequestDto;
import webcrawler.urldiscovery.response.UrlDiscoveryErrorDto;
import webcrawler.urldiscovery.response.UrlDiscoveryResponseDto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.ArrayList;

@Slf4j
@Service
@RequiredArgsConstructor
public class UrlDiscoveryService {
    private final UrlDiscoveryEventProducer urlDiscoveryEventProducer;

    public UrlDiscoveryResponseDto createUrlDiscovery(UrlDiscoveryRequestDto requestDto) {
        try {
            return urlDiscoveryEventProducer.publishUrlDiscoveryCreate(requestDto);

        } catch (Exception e) {
            log.error("[UrlDiscoveryService.createUrlDiscovery] url={}", requestDto.getStartUrl(), e);
            throw e;
        }
    }

    public UrlDiscoveryBatchResponseDto createUrlDiscoveries(List<UrlDiscoveryRequestDto> requestDtos) {
        List<UrlDiscoveryResponseDto> successResults = new ArrayList<>();
        List<UrlDiscoveryErrorDto> errors = new ArrayList<>();

        for (UrlDiscoveryRequestDto requestDto : requestDtos) {
            try {
                UrlDiscoveryResponseDto responseDto = urlDiscoveryEventProducer.publishUrlDiscoveryCreate(requestDto);
                successResults.add(responseDto);

            } catch (Exception e) {
                log.error("[UrlDiscoveryService.createUrlDiscoveries] url={}", requestDto.getStartUrl(), e);

                errors.add(UrlDiscoveryErrorDto.create(requestDto.getStartUrl(), e.getMessage()));
            }
        }

        if (successResults.isEmpty()) {
            throw new RuntimeException("All URL discovery requests failed");
        }

        return UrlDiscoveryBatchResponseDto.create(successResults, errors);
    }
}

File: sdip09-web-crawler-ktor/web-crawler-app/src/main/kotlin/webcrawler/domain/urldiscovery/UrlDiscoveryService.kt
package webcrawler.domain.urldiscovery

import org.slf4j.LoggerFactory

class UrlDiscoveryService(
    private val producer: UrlDiscoveryProducer
) {
    private val log = LoggerFactory.getLogger(UrlDiscoveryService::class.java)

    suspend fun createUrlDiscovery(requestDto: UrlDiscoveryRequestDto): UrlDiscoveryResponseDto {
        return try {
            producer.publishUrlDiscoveryCreate(requestDto)
        } catch (e: Exception) {
            log.error("[UrlDiscoveryService] createUrlDiscovery failed: url={}", requestDto.startUrl, e)
            throw e
        }
    }

    suspend fun createUrlDiscoveries(requestDtos: List<UrlDiscoveryRequestDto>): UrlDiscoveryBatchResponseDto {
        val successResults = mutableListOf<UrlDiscoveryResponseDto>()
        val errors = mutableListOf<UrlDiscoveryErrorDto>()

        for (requestDto in requestDtos) {
            try {
                val response = producer.publishUrlDiscoveryCreate(requestDto)
                successResults.add(response)
            } catch (e: Exception) {
                log.error("[UrlDiscoveryService] createUrlDiscoveries failed: url={}", requestDto.startUrl, e)
                errors.add(UrlDiscoveryErrorDto.create(requestDto.startUrl, e.message ?: "Unknown error"))
            }
        }

        if (successResults.isEmpty()) {
            throw RuntimeException("All URL discovery requests failed")
        }

        return UrlDiscoveryBatchResponseDto.create(successResults, errors)
    }
}
